Add support of run time flags for enable Castanets

santhoshkerur opened the pull request:
This change adds runtime flag to enable castanets.
Runtime flags --enable-forking, --server-address has been removed.
Castanets can be verified as below,

Standalone mode:
chrome <url>

Castanets mode:
Browser: chrome <url> --enable-castanets
Renderer: chrome --type=renderer --enable-castanets=<IP Address>

Signed-off-by: Santhoshkumar Kerur <santhosh.vk@samsung.com>
Signed-off-by: suyambu.rm <suyambu.rm@samsung.com>

suyambulingamrm commented on components/discardable_memory/client/client_discardable_shared_memory_manager.cc:
```diff
@@ -360,33 +362,37 @@ ClientDiscardableSharedMemoryManager::AllocateLockedDiscardableSharedMemory(
+    }
+    //*shared_memory_region = memory->DuplicateRegion();
+    // Close file descriptor to avoid running out.
+    // memory->Close(); needed ?
```
May be we can remove these commented lines too.

santhoshkerur replied:
okay.

suyambulingamrm commented on content/browser/renderer_host/render_process_host_impl.cc:
```diff
@@ -3105,7 +3109,8 @@ void RenderProcessHostImpl::PropagateBrowserCommandLineToRenderer(
 #if defined(CASTANETS)
-    switches::kServerAddress,
+    switches::kEnableCastanets,
+   //switches::kServerAddress,
```
can remove commented lines

santhoshkerur replied:
okay. Done.

suyambulingamrm reviewed (commented):
There are many changes specific to ANDROID and CASTANETS in below files..
We need to cross check if runtime checks are needed for those too.

gpu/ipc/common/gpu_memory_buffer_support.cc
gpu/ipc/service/gpu_memory_buffer_factory.cc
ui/gfx/mojo/buffer_types_struct_traits.cc
content/browser/frame_host/frame_tree_node.cc
content/browser/frame_host/render_frame_host_impl.cc	
content/browser/android/content_startup_flags.cc
content/browser/child_process_launcher_helper_android.cc
content/public/browser/content_browser_client.cc
content/public/common/use_zoom_for_dsf_policy.cc
content/public/common/content_features.cc
content/renderer/render_widget.cc
content/renderer/input/render_widget_input_handler.cc
third_party/blink/renderer/core/loader/resource/script_resource.cc
third_party/blink/public/common/page/launching_process_state.h
mojo/public/cpp/system/platform_handle.cc
mojo/core/platform_handle_utils.cc
mojo/core/broker_castanets.cc
mojo/core/core.cc
mojo/core/broker_host.cc
mojo/core/shared_buffer_dispatcher.cc

santhoshkerur commented:
> There are many changes specific to ANDROID and CASTANETS in below files..
> We need to cross check if runtime checks are needed for those too.
> 
> gpu/ipc/common/gpu_memory_buffer_support.cc
> gpu/ipc/service/gpu_memory_buffer_factory.cc
> ui/gfx/mojo/buffer_types_struct_traits.cc
> content/browser/frame_host/frame_tree_node.cc
> content/browser/frame_host/render_frame_host_impl.cc
> content/browser/android/content_startup_flags.cc
> content/browser/child_process_launcher_helper_android.cc
> content/public/browser/content_browser_client.cc
> content/public/common/use_zoom_for_dsf_policy.cc
> content/public/common/content_features.cc
> content/renderer/render_widget.cc
> content/renderer/input/render_widget_input_handler.cc
> third_party/blink/renderer/core/loader/resource/script_resource.cc
> third_party/blink/public/common/page/launching_process_state.h
> mojo/public/cpp/system/platform_handle.cc
> mojo/core/platform_handle_utils.cc
> mojo/core/broker_castanets.cc
> mojo/core/core.cc
> mojo/core/broker_host.cc
> mojo/core/shared_buffer_dispatcher.cc

@suyambulingamrm , Files have changed. Thanks for review and Support as co-contributor.

suyambulingamrm commented:
+1.
Verified on desktop, android.
Below are not part of this patch yet.
1. Minimise castanets main activity only when --enable-castanets is provided. (Inprogress will be part of new patchset)
2.  We were using BGRA as Skia format for castanets to make android renderer compatible with chrome/efl browser. Because of this we see incorrect colors on some pages, when we run android browser without runtime flag. (TODO)

Sunwoo-Nam commented on components/services/font/public/cpp/font_service_thread.cc:
```diff
@@ -144,8 +144,25 @@ scoped_refptr<MappedFontFile> FontServiceThread::OpenStream(
+      // The font-service may have been killed.
+      return nullptr;
+    }
+  }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on cc/raster/staging_buffer_pool.cc:
```diff
@@ -261,7 +266,15 @@ std::unique_ptr<StagingBuffer> StagingBufferPool::AcquireStagingBuffer(
+      if (!busy_buffers_.front()->query_id ||
+          !CheckForQueryResult(ri, busy_buffers_.front()->query_id))
+        break;
+    }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on content/renderer/render_widget.cc:
```diff
@@ -1664,7 +1668,11 @@ void RenderWidget::IntrinsicSizingInfoChanged(
+      Send(new WidgetHostMsg_DidFirstVisuallyNonEmptyPaint(routing_id_));
+    else
+      QueueMessage(
+          new WidgetHostMsg_DidFirstVisuallyNonEmptyPaint(routing_id_));
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on gpu/command_buffer/client/implementation_base.cc:
```diff
@@ -261,16 +265,20 @@ gpu::ContextResult ImplementationBase::Initialize(
+    mojo::WaitSyncSharedMemory(transfer_buffer_->shared_memory_guid());
+  } else {
+    helper_->Finish();
+  }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/core/broker_host.cc:
```diff
@@ -74,10 +78,15 @@ bool BrokerHost::SendChannel(PlatformHandle handle) {
+    data->port = -1;
+  } else {
+    message = CreateBrokerMessage(BrokerMessageType::INIT, 1, nullptr);
+  }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/core/channel_posix.cc:
```diff
@@ -504,13 +506,18 @@ class ChannelPosix : public Channel,
+        }
+      } else {
+        AcceptSocketConnection(server_.platform_handle().GetFD().get(),
+                               &socket_);
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/core/node_controller.cc:
```diff
@@ -250,17 +253,32 @@ void NodeController::AcceptBrokerClientInvitation(
+      CancelPendingPortMerges();
+      return;
+    }
+    connection_params = ConnectionParams(std::move(endpoint));
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/core/node_controller.cc:
```diff
@@ -487,17 +516,31 @@ void NodeController::SendBrokerClientInvitationOnIOThread(
+                       process_error_callback);
     channel_ok = broker_host->SendChannel(
         node_channel.TakeRemoteEndpoint().TakePlatformHandle());
   }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/core/node_controller.cc:
```diff
@@ -1299,36 +1350,42 @@ void NodeController::OnIntroduce(const ports::NodeName& from_node,
+      DVLOG(1) << "Could not be introduced to peer " << name;
+      base::AutoLock lock(peers_lock_);
+      pending_peer_messages_.erase(name);
     }
```
Is it possible to remove duplicate code?

Sunwoo-Nam commented on mojo/public/cpp/system/invitation.cc:
```diff
@@ -97,17 +101,24 @@ void SendInvitation(ScopedInvitationHandle invitation,
+    result =
+        MojoSendInvitation(invitation.get().value(), &process_handle, &endpoint,
+                           error_handler, error_handler_context, &options);
+  }
```
Is it possible to remove duplicate code?

Sunwoo-Nam reviewed (commented):
I commented about duplicate codes, but it is not critical.

Commit:
Add support of run time flags for enable Castanets

This change adds runtime flag to enable castanets.
Runtime flags --enable-forking, --server-address has been removed.
Castanets can be verified as below,

Standalone mode:
chrome <url>

Castanets mode:
Browser: chrome <url> --enable-castanets
Renderer: chrome --type=renderer --enable-castanets=<IP Address>

Signed-off-by: suyambu.rm <suyambu.rm@samsung.com>
Signed-off-by: Santhoshkumar Kerur <santhosh.vk@samsung.com>

suyambulingamrm commented:
> I commented about duplicate codes, but it is not critical.

Done.

Sunwoo-Nam reviewed (approved):
Looks good to me.
